### PR TITLE
feat: a simple approach to allow keyboard navigation for menu items

### PR DESCRIPTION
**Pull request by ctron**
Also: provide some helpers to work with HTML stuff

### CODE REVIEW

File: src/components/menu/mod.rs [hunk 1]
```diff
@@ -69,9 +73,111 @@ pub(crate) struct MenuListProperties {
...
+                }
+            })
+            // or default
+            .unwrap_or_else(|| if offset > 0 { 0 } else { len.saturating_sub(1) });
+
+        // get as node
+        let next_node = items
+            .get(next_index)
+            .and_then(focusable_element)
+            .and_then(|ele| ele.dyn_into::<HtmlElement>().ok());
```
Comment (aDogCalledSpot) [inline src/components/menu/mod.rs:171]: let Some else should be nicer here

File: src/utils/html.rs [hunk 2]
```diff
@@ -18,3 +17,102 @@ pub fn value(node_ref: &NodeRef) -> Option<String> {
...
+}
+
+impl ElementSupport for NodeRef {
+    fn contains(&self, target: Option<web_sys::EventTarget>) -> bool {
+        let target = target.as_ref().and_then(|target| target.dyn_ref::<Node>());
+        if let Some(element) = self.cast::<Element>() {
+            element.contains(target)
+        } else {
+            false
+        }
```
Comment (aDogCalledSpot) [inline src/utils/html.rs:34]: `self.cast::<Element>().is_some_and(|element| element.contains(target))`

**Review by aDogCalledSpot — CHANGES_REQUESTED** [review]
Really cool stuff! Only have some really minor stylistic things in the review.

Could this be extracted to a utilities file? This can be used for things other than menus as well (DualListSelector is the first one that comes to mind).

It will probably end up being exactly the same thing as [patternfly-react's keyboard handler](https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/helpers/KeyboardHandler.tsx).

File: src/components/menu/mod.rs [hunk 3]
```diff
@@ -69,9 +73,111 @@ pub(crate) struct MenuListProperties {
...
         </ul>
     )
 }
+
+fn focusable_element(element: &HtmlElement) -> Option<HtmlElement> {
+    element
+        .query_selector("a, button, input")
+        .ok()
+        .flatten()
+        .and_then(|element| element.dyn_into::<HtmlElement>().ok())
```
[the remark this answers is not in the document]
Comment (ctron) [reply src/components/menu/mod.rs:101]: Yea, I forgot rust can do that! :grin:

**Review by aDogCalledSpot — APPROVED** [review]

**Pull request merged**